### ISSUE TITLE
SPFarm without PsRunAsCredential 

### ISSUE THREAD

**Issue by Harish9211** [PowerShell/SharePointDsc#833]
Is it possible to run SpFarm and other resources without PsRunAsCredential, as i dont want to provide creds on script?

I am using Pull mode so i cant use Get-Credential

**Comment by ykuijs** [comment]
You have to use either the InstallAccount or PsRunAsCredential parameter.

You are either pushing or pulling a configuration MOF file. The credentials are not required during a push or pull, but during the compilation phase (creation of the MOF file) and are then included into the MOF file. So you don't need Get-Credential during the pull or push.

**Comment by Harish9211** [comment]
Yes so i have to provide credentials in the script as plain text which i dont want to do. Is there any way i can make this script to pick my logged in account and run all task with that account?

**Comment by johlju** [comment]
@Harish9211 You have to compile the configuration to contain the credentials to use when LCM runs. LCM will decrypt the credentials (using the certificate you setup). See here for more information https://docs.microsoft.com/sv-se/powershell/dsc/pullserver

It's not possible to assign credentials when LCM runs, they need to already be in the configuration.

You can add a parameter to your configuration to pass in the credential, see the example here https://github.com/PowerShell/SharePointDsc/blob/dev/Modules/SharePointDsc/Examples/Resources/SPFarm/1-CreateBasicFarm.ps1
The credential will be encrypted using the certificate you provide, see https://docs.microsoft.com/en-us/powershell/dsc/securemof

**Comment by ykuijs** [comment]
@Harish9211 The first link is the Swedish version......the English version can be found at: https://docs.microsoft.com/en-us/powershell/dsc/pullserver :-)

In short: The process of using DSC is:
1. Create a DSC Configuration in a PS1 file. Since this is PowerShell based, you can use PowerShell commands to make it dynamic.
2. Create a MOF file by running the script
3. Push the MOF file to the server or have the server pull the MOF file from a pull server

In the script you have to use credentials to specify the various accounts needed for SharePoint. But only when you generate the MOF file (step 2) you need to enter the credentials, either by
1. Including them into the script as plain text: Not recommended, security issue
2. Using encrypted credentials: Can only be decrypted by the user that encrypted the password on the server it was encrypted on.
3. Entering them manually: Since this is an activity you don't perform often, I usually see this option being used.

The MOF generation process has two options:
1. Store the credentials in plain text into the MOF file. NOTE: This is a security issue, only for testing, absolutely not recommended for production use. You have to use the property "PSDscAllowPlainTextPassword = $true" to allow passwords to be stored as plain text: https://blogs.technet.microsoft.com/ashleymcglone/2015/12/18/using-credentials-with-psdscallowplaintextpassword-and-psdscallowdomainuser-in-powershell-dsc-configuration-data/
2. Use certificates to encrypt your passwords, see the SecureMOF link @johlju shared in the previous message.

**Comment by ykuijs** [comment]
@Harish9211 Is your question answered?
